src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs

```diff
@@ -193,30 +224,81 @@ private NamedTypeSymbol ConstructAnonymousDelegateImplementationSymbol(Anonymous
             var typeDescr = anonymous.TypeDescriptor;
             Debug.Assert(typeDescr.Location.IsInSource); // AnonymousDelegateTemplateSymbol requires a location in source for ordering.
 
-            var fields = typeDescr.Fields;
-            bool returnsVoid = fields.Last().Type.IsVoidType();
-            int nTypeArguments = fields.Length - (returnsVoid ? 1 : 0);
-            var refKinds = default(RefKindVector);
-            if (fields.Any(f => f.RefKind != RefKind.None))
+            var typeParameters = anonymous.ContainingTypeParameters;
+
+            // If all parameter types and return type are valid type arguments, construct
+            // the delegate type from a generic SynthesizedDelegateSymbol template.
+            // Otherwise, use a non-generic AnonymousDelegateTemplateSymbol.
+            if (!allValidTypeArguments(typeDescr))
```

Comment:
>if (!allValidTypeArguments(typeDescr)) [](http://example.com/codeflow?start=12&length=38)

I do not think this split is justified. We should be able to create the right template without introducing the whole new way of managing templates. The templates can be slightly different, but they are all can be managed the same way, I think.